cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,323 @@
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			return err
+		}
+		if currentFirewall == nil {
+			klog.Infof("an existing firewall not found, we need to create one")
+			currentFirewall, err = fm.createFirewall(ctx, svcInboundRules)
+			if err != nil {
+				return err
+			}
+			klog.Infof("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, svcInboundRules []godo.InboundRule) (*godo.Firewall, error) {
+	fr := fm.createFirewallRequest(fm.workerFirewallName, fm.workerFirewallTags, svcInboundRules)
+	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
+	return currentFirewall, err
+}
+
+func (fm *firewallManagerOp) createFirewallRequest(fwName string, fwTags []string, rules []godo.InboundRule) *godo.FirewallRequest {
+	return &godo.FirewallRequest{
+		Name:         fwName,
+		InboundRules: rules,
+		Tags:         fwTags,
+	}
+}
+
+func (fc *firewallCache) updateCache(currentFirewall *godo.Firewall) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.firewall = currentFirewall
+}
+
+func (fc *FirewallController) ensureReconciledFirewall(ctx context.Context) error {
+	serviceList, err := fc.serviceLister.List(labels.Everything())
+	if err != nil {
+		return errFailedToListServices
+	}
+	inboundRules := fc.createInboundRules(serviceList)
+	fc.fwManager.Set(ctx, inboundRules)
+	return nil
+}
+
+func (fc *FirewallController) createInboundRules(serviceList []*v1.Service) []godo.InboundRule {
+	var nodePortInboundRules []godo.InboundRule
+	for _, svc := range serviceList {
+		if svc.Spec.Type == v1.ServiceTypeNodePort {
+			// this is a nodeport service so we should check for existing inbound rules on all ports.
+			for _, servicePort := range svc.Spec.Ports {
+				// In the odd case that a failure is asynchronous causing the NodePort to be set to zero.
+				if servicePort.NodePort == 0 {
+					klog.Warningf("NodePort on the service is set to zero")
+					continue
+				}
+				nodePortInboundRules = append(nodePortInboundRules,
+					godo.InboundRule{
+						Protocol:  "tcp",
+						PortRange: strconv.Itoa(int(servicePort.NodePort)),
+						Sources: &godo.Sources{
+							Tags: fc.workerFirewallTags,
+						},
+					},
+					godo.InboundRule{
+						Protocol:  "udp",
+						PortRange: strconv.Itoa(int(servicePort.NodePort)),
+						Sources: &godo.Sources{
+							Tags: fc.workerFirewallTags,
+						},
+					},
+					godo.InboundRule{
+						Protocol:  "icmp",
+						PortRange: strconv.Itoa(int(servicePort.NodePort)),
+						Sources: &godo.Sources{
+							Tags: fc.workerFirewallTags,
+						},
+					},
+				)
+			}
+		}
+	}
+	return nodePortInboundRules
+}
+
+func (fc *firewallCache) isEqual(fw *godo.Firewall) bool {
+	fc.mu.RLock()
+	defer fc.mu.RUnlock()
+	return cmp.Equal(fc.firewall, fw)
+}
+
+func (fc *firewallCache) setEqualToCachedFirewall() *godo.Firewall {
```

Comment:
Do we need this method? From a quick glance, it looks like every time we invoke it we do so to either check for equality or retrieve the firewall ID. For both cases, we have dedicated methods (`isEqual` and `getFirewallID`).

Or am I missing something?

Reply:
If we do need this function, can we rename it to `getCachedFirewall()` or similar? The `set` in the name makes it sound like it updates the cache, but that's clearly not the intent.

Reply:
@timoreimann 

context in regards to where these were invoked:

```go
fw := fm.fwCache.setEqualToCachedFirewall()
	if fw != nil {
```

_and_

```go
targetFirewall := fm.fwCache.setEqualToCachedFirewall()
	if targetFirewall != nil ... {
```

In both cases I am using this method to set a variable to be the value of the cached firewall. Since the goal with `isEqual` and `getFirewallID` was to get rid of the need to deal with locking in various places, I needed something to address the final 2 cases that didn't exactly just need a `check of firewall equality` or just to `get a firewall ID`.

The alternative is this:
```go
fm.fwCache.mu.RLock()
fw := fm.fwCache.firewall
fm.fwCache.mu.RUnlock()
```

To sum it up, there were 3 locking cases. Not just 2 (unless I'm mistaken or not taking advantage of some potential optimization?):
**Case 1:** _Is the current firewall equal to the cached firewall?_
 `cmp.Equal(fm.fwCache.firewall, currentFw)`
**Case 2:** _We need to check for nil, so create a variable to store the current cached representation of a firewall._
`fw := fm.fwCache.firewall`
**Case 3:** _Let's actually update the cache itself._
`fm.fwCache.firewall = currentFw`

Reply:
I was preparing a long-ish response first explaining how we always use `getFirewallID()` after a call to `getCachedFirewall()` (née `setEqualToCachedFirewall`) and argued we could define the semantics for the returned firewall ID such that 0 meant "no firewall cached" in order to remove `getCachedFirewall()`. However, the other use case we have is that we [need to compare the inbound rules to the target firewall's inbound rules](https://github.com/digitalocean/digitalocean-cloud-controller-manager/pull/332/files#diff-92ac4af525e02d85b1ee8eb33e17ad3bR184), which we cannot eliminate so easily. At least, none of the options I came up with in my head made it look more elegant overall.

If that is the case though, then I think we don't really need `getFirewallID()` since we can just access the `ID` field on the firewall we get from `getCachedFirewall()`. WDYT?

Reply:
I think this makes sense, good idea!